<a href="https://colab.research.google.com/github/dakyungsilvialee/Financial-Accounting-with-Python/blob/main/Analyzing_Australian_Tax_Return_Data_Part_I_2022.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Analyzing Australian Tax Return Data- Part 1

## Background Information

In this case study you will explore Australian corporations’ tax return data. Australia (unlike the United States) publicly discloses three line items from large corporations’ tax returns. The three line items disclosed are:

- **Total Income**- Total income is a measure of a company’s gross receipts or revenue. It does not include any deductions or expenses.


- **Taxable Income** - Taxable income is a measure of a company’s profitability based on Australian tax law. Taxable income is the basis from which a company’s income tax liability is calculated.
   - The difference between a company’s total income and it’s taxable income is the amount of tax deductions or expenses a company claims on its tax return.
   
<center> <font color = blue> <font size = 4> $Taxable Income$ <font color = black> $=$ <font color = red> $Total  Income$ <font color = black> $- Tax Deductions$ </center>

   - Taxable income is separate and distinct from a company’s financial accounting income. Although both financial accounting income and taxable income measure a firm’s profitability, they are calculated differently.
      - Financial accounting income is designed to measure a firm’s economic transactions and to be as informative as possible to investors and creditors.
      - Taxable income is defined by the government and is designed to raise money and to encourage or discourage certain behaviors. For example, in the United States, we are allowed to deduct interest incurred on a household mortgage because the government wants to encourage home ownership.


- **Tax Payable**- Tax payable is a company’s Australian income tax liability and measures the amount of income taxes that the company must pay to Australia’s federal government.
   - Tax payable is a function of a company’s taxable income. If the company does not claim any tax credits, tax payable will equal 30% of a company’s taxable income. 30% is the corporate income tax rate in Australia. 
   
<center> <font color = green> <font size = 4> $Tax Payable$ <font color = black> $=$ <font color = blue> $Taxable Income$ <font color = black> $*Tax Rate (30\%\ in Australia)$ </center>

   - Tax payable will be less than a company’s taxable income times the income tax rate if the company claims any tax credits.

<center> <font color = green> <font size = 4> $Tax Payable$ <font color = black> $=$ <font color = blue> $Taxable Income$ <font color = black> $*30\%-Tax Credits$ </center>
	
   - Tax credits are a way for the government to subsidize certain corporate activities. Tax credits allow the government to indirectly pay a corporation back for the amount of money a corporation spent on a certain activity. For instance, in the United States, some students qualify for tuition tax credits. By giving a student a tuition tax credit, the government allows a student to pay less taxes than the amount the student would normally owe based on the student’s taxable income. This way, the student can pay for the student’s tuition with the money the student saved by not paying taxes. These transactions are an indirect way for the government to pay for the student’s tuition. In corporations, the most common tax credit is for research and development (R&D). Many governments want to encourage R&D as innovation often leads to economic growth.

The following equation summarizes the relationship between the three variables:

<center> <font size = 4> <font color = red> $(Total Income$ </font> $-Tax Deductions)*30\%-Tax Credits=$ <font color = green> $Tax Payable$ </font> </center>

The Australian government started to publicly disclose large corporation’s tax return information in December 2015. Due to the long nature of the tax reporting cycle, information disclosed in December 2015 relates to tax returns for years ending December 1, 2013 – November 30th, 2014 (most Australian firms have a June 30th year end).

The Australian government’s tax transparency legislation has two stated objectives: 

> 1. Discourage large corporate entities from engaging in aggressive tax avoidance practices
> 2. Provide more information to inform public debate about tax policy
> *Australian Senate 2013* 
    
    
Australian legislators hoped that if large firms were forced to report their tax information, public shaming would cause low tax paying firms (firms that engage in a lot of tax planning) to pay more taxes. In this case study, you will explore that idea.

## Case Instructions

To break down the over arching question of this case into actionable steps, you will follow the FACT framework. 

## F- Frame the Question: 

The overarching goal of this case is to provide evidence as to whether public disclosure of corporate tax return data affected firms’ tax planning choices. Remember that Australian legislators hoped that if large firms were forced to report their tax return information, the public would shame low tax paying firms (firms that engage in a lot of tax planning) and subsequently cause these firms to pay more taxes. We are specifically interested in measuring if this phenomenon occurred. 

Because you are studying income taxes, it is important to remember that taxes are a function of profitability. Rather than focusing on total taxes paid in isolation, you may want to focus on an effective tax rate:

<div align = "center"> <font size = 4> $Effective Tax Rate = \frac{Tax Payable}{Taxable Income}$ </font> </div>

If this number is lower than Australia’s 30% statutory rate, the firm is claiming tax credits.

Keep in mind that a firm that pays the 30% statutory rate may still engage in tax planning by lowering the firm’s taxable income. With this type of tax planning, a firm claims extra tax deductions. Some tax deductions are normal expenses, such as an employee's salary. Some tax deductions may push the boundaries of what lawmakers intended corporations to deduct. For instance, a firm may improperly deduct the cost of operating a corporate jet for an executive’s vacation.

As an example of tax planning that relies on claiming extra tax deductions, a recent article in *The New York Times* notes that,
>...without any explanation in his [President Donald Trump] returns, the general and administrative expenses at his Bedminster golf club in New Jersey increased fivefold from 2016 to 2017.
>
>*Buettner, Craig and McIntire 2020*

We have panel data- multiple firms over multiple years. The law was passed in June 2013. The first data was disclosed in December 2015 and related to tax years ending 12/1/2013 - 11/30/2014; most Australian corporations have a June 30th year end so most of the data related to the year ending June 30th, 2014. Companies could have updated their tax planning for this year in anticipation of a public backlash but the first two years, the year ending 6/30/2014 and 6/30/2015 ended before the first year of data was disclosed in December 2015. If companies changed their tax planning after an actual public backlash (if there was one) we would not see it in the data until year 3 ending 6/30/2016 (which was half way over when the first year of data was disclosed) or even year 4 ending 6/30/2017 (which occured completely after data had been disclosed).

**Keeping in mind the above points and your understanding of the data, <font color = 'blue'> write three concrete data questions </font> that are designed to help you understand (at least descriptively) if public tax return disclosure led firms to pay more taxes.**

A few notes:
- Your questions should be framed in terms of the data you have and be very specific. For instance, if you are interested in an average, specify whether you will measure the mean or median. Note that the same question augmented to ask about the mean instead of the median does **not** count as a second question.
- Your question should specify which year you are looking at: first year, last year, changes over specific years.
- You may want to study all firms, or you may want to study specific subgroups (i.e. public vs. private, really big vs. less big (define size criteria), etc.)
- Be sure that your questions can be answered with the data we have. If you have a question that cannot be answered with the data you have, describe what data you need and how you would use it to help answer your question.
- Your questions should help to answer the bigger question as to whether public tax return disclosure changed tax paying behavior.
- I am looking for thoughtful ideas and most importantly, **I should be able to understand exactly what you intend to measure.**

**Good and not as good examples based on a study of Lyft and Uber pricing strategies:**
- Example of an interesting question that does not clearly explain what will be measured
> We expect the surge multiplier to increase at certain times of day such as rush hour. Is this expectation true? What is the relationship between surge multiplier and time?

- Same question, more specifics with respect to what will be measured
> What is the highest surge multiplier (from both Lyft and Uber) observed in our dataset and what time of day does it occur? If I segment the data into 4 times of day (7am – 9am, 10am-4pm, 5pm- 7pm, 7pm-7am), what is the average (mean and median) surge multiplier during each time segment? 

<div style="color:black;
           display:fill;
           border-radius:5px;
           background-color:#ABBAEA;>

            
<p style="padding: 10px;
              color:white;">

**Type your specific, measurable questions here**:


. If I separate public and private firms observed in our dataset and segment the years (2013-2014, 2014-2015, 2015-2016, 2016-2017, 2017-2018, 2018-2019, 2019-2020), **what is the mean and standard deviation of the effective tax rate for those periods?** Thus, what would you conclude from the clustering of firms (public vs private) such as if the policy is doing its job or if firms are following the same/different tendencies? 


. After determining the size of the firm- based if under or above the median of total income- **what is the average (median) difference between total income and taxable income in years 2013-2014 and 2014-2015** in order to see whether big/small companies take more/less advantage of tax credits and deductions.


. **What is the average change of the effective tax rate between years(2013-2014, 2014-2015)?** By separating public and private firms, how can we conclude that a company is ashamed/proud of disclosing their tax return data?


------

### A- Assemble the Data- Part 1

The Australian Tax Office (ATO) is the Australian version of the United States’ Internal Revenue Service (IRS). The ATO posts excel workbooks with large corporations' tax return information once a year on the following [website](https://data.gov.au/dataset/ds-dga-c2524c87-cea4-4636-acac-599a82048a26/details). 

Each workbook relates to a different year of data. You will analyze the first seven years of data which relate to tax years ending December 1st, 2013- November 30th, 2020. The data is a time series with seven years of observations for approximately 1,000 firms.


**1. Open the first excel workbook in excel. Read the information tab in the excel workbooks to better understand the data.**
   - In the first year of public tax return disclosure, public firms' data was disclosed in December, while private firms' data was disclosed in March. In all subsequent years, public and private firm data is disclosed at the same time (in December). Public and private firms are separated only in the first year of data: public firms on the tab labeled ‘December’ and private firms on the tab labeled ‘March’. In all subsequent years, the two sets of firms are intermingled in the data. 
   - You should also notice that in all years after the first year, there are additional tabs or rows disclosing prior year data. You will only focus on data that is filed on time so you can ignore those tabs (for instance the 2013-2014 data in the 2014-2015 workbook).

   
    
**2. Without doing any manipulation in excel, upload the data you need from each excel workbook into this Jupyter notebook and clean the data <font color = 'blue'> using a loop </font>. Be sure to document your work and to make it as replicable as possible.**
- I recommend uploading the first year of data outside of the loop for two reasons:
   - The first year of data is different because public and private data are listed on separate tabs. You do want to identify and distinguish public and private firms. Use the first year of data to do so. We will assume no public firms went private and vice versa during our time frame.
   - Uploading and cleaning this data separately will make writing a loop easier for all subsequent years
   - **Things to keep in mind when uploading and cleaning:**
      - When uploading the data, keep in mind that you are interested specifically in income tax data, not PRRT (Petroleum Resource Rent Tax). 
      - Do not keep any firms that are missing their Australian Business Number (ABN).
      - Make sure each year of data is distinguishable, i.e. create a column to indicate which year is which
      - All public firms with total income over \\$100M are subject to disclosure whereas private firms are only subject to disclosure if the firm's total income is over \\$200M. To make the two sets of firms fully comparable, filter each year of data to only include firms with \\$200M or more in total income. 


       
<font color= 'green'> *If you are really struggling with the loop. Upload each file separately and move on. You will lose some points for not having the loop but it is more important that you finish this notebook so that you can move on to part 2- the data analysis.*


In [3]:
!pip install xlrd
!pip install psycopg2

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [4]:
import numpy as np
from google.colab import files
import io
import pandas as pd

pd.options.display.float_format = '{:,.2f}'.format

#import midter dataset files
dataset = files.upload()

Saving 2013-14-corporate-report-of-entity-tax-information.xlsx to 2013-14-corporate-report-of-entity-tax-information (1).xlsx
Saving 2014-15-corporate-report-of-entity-tax-information.xlsx to 2014-15-corporate-report-of-entity-tax-information (1).xlsx
Saving 2015-16-corporate-report-of-entity-tax-information.xlsx to 2015-16-corporate-report-of-entity-tax-information (1).xlsx
Saving 2016-17-corporate-report-of-entity-tax-information.xlsx to 2016-17-corporate-report-of-entity-tax-information (1).xlsx
Saving 2017-18-corporate-report-of-entity-tax-information.xlsx to 2017-18-corporate-report-of-entity-tax-information (1).xlsx
Saving 2018-19-corporate-report-of-entity-tax-information.xlsx to 2018-19-corporate-report-of-entity-tax-information (1).xlsx
Saving 2019-20-corporate-report-of-entity-tax-information.xlsx to 2019-20-corporate-report-of-entity-tax-information (1).xlsx


In [5]:
#first year public dataset
df_public = pd.read_excel('2013-14-corporate-report-of-entity-tax-information.xlsx', sheet_name = 'December')
df_public

Name               ABN  \
0                            3M AUSTRALIA PTY LTD 90,000,100,096.00   
1                              A P EAGERS LIMITED 87,009,680,013.00   
2               A2 AUSTRALIAN INVESTMENTS PTY LTD 93,126,014,275.00   
3                                    AAPC LIMITED 87,009,175,820.00   
4                   ABACUS GROUP HOLDINGS LIMITED 31,080,604,619.00   
...                                           ...               ...   
1533          YUSEN LOGISTICS (AUSTRALIA) PTY LTD 85,003,550,776.00   
1534                ZF HOLDINGS AUSTRALIA PTY LTD 85,109,711,828.00   
1535             ZIMMER AUSTRALIA HOLDING PTY LTD 93,107,449,534.00   
1536                 ZOETIS AUSTRALIA PTY LIMITED 94,156,476,425.00   
1537  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   

      Total income $  Taxable income $  Tax payable $  
0          452892659     86,587,682.00  25,328,872.00  
1         2561960532     91,824,776.00  22,270,429.00  
2          110173384        783,187.00     234,956.00  
3          570890761     49,107,218.00  13,273,263.00  
4          198340272     19,042,211.00   5,712,663.00  
...              ...               ...            ...  
1533       111253578      2,564,125.00     769,238.00  
1534       204210113     18,799,144.00   4,662,674.00  
1535       188019618     39,265,645.00  11,779,694.00  
1536       256720802     16,509,304.00   4,952,791.00  
1537      2449944648    224,490,384.00  47,810,511.00  

[1538 rows x 5 columns]

In [6]:
#drop ABN null value

df_public = df_public.dropna(subset=['ABN'])
df_public 

Name               ABN  \
0                            3M AUSTRALIA PTY LTD 90,000,100,096.00   
1                              A P EAGERS LIMITED 87,009,680,013.00   
2               A2 AUSTRALIAN INVESTMENTS PTY LTD 93,126,014,275.00   
3                                    AAPC LIMITED 87,009,175,820.00   
4                   ABACUS GROUP HOLDINGS LIMITED 31,080,604,619.00   
...                                           ...               ...   
1533          YUSEN LOGISTICS (AUSTRALIA) PTY LTD 85,003,550,776.00   
1534                ZF HOLDINGS AUSTRALIA PTY LTD 85,109,711,828.00   
1535             ZIMMER AUSTRALIA HOLDING PTY LTD 93,107,449,534.00   
1536                 ZOETIS AUSTRALIA PTY LIMITED 94,156,476,425.00   
1537  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   

      Total income $  Taxable income $  Tax payable $  
0          452892659     86,587,682.00  25,328,872.00  
1         2561960532     91,824,776.00  22,270,429.00  
2          110173384        783,187.00     234,956.00  
3          570890761     49,107,218.00  13,273,263.00  
4          198340272     19,042,211.00   5,712,663.00  
...              ...               ...            ...  
1533       111253578      2,564,125.00     769,238.00  
1534       204210113     18,799,144.00   4,662,674.00  
1535       188019618     39,265,645.00  11,779,694.00  
1536       256720802     16,509,304.00   4,952,791.00  
1537      2449944648    224,490,384.00  47,810,511.00  

[1531 rows x 5 columns]

In [7]:
#filter each year of data to only include firms with $200M or more in total income

df_public = df_public[df_public['Total income $'] >= 200000000]
df_public.head(20)

Name               ABN  \
0                          3M AUSTRALIA PTY LTD 90,000,100,096.00   
1                            A P EAGERS LIMITED 87,009,680,013.00   
3                                  AAPC LIMITED 87,009,175,820.00   
5       ABB GROUP INVESTMENT MANAGEMENT PTY LTD 47,082,803,852.00   
6                ABBOTT AUSTRALASIA PTY LIMITED 95,000,180,389.00   
7                            ABBVIE PTY LIMITED 48,156,384,262.00   
9                       ABP FINANCE PTY LIMITED 58,139,991,101.00   
11         ACCENTURE AUSTRALIA HOLDINGS PTY LTD 61,096,995,649.00   
12             ACCESS PREPAID AUSTRALIA PTY LTD 47,145,452,044.00   
15           ACE AUSTRALIA HOLDINGS PTY LIMITED 64,116,987,618.00   
16          ACER COMPUTER AUSTRALIA PTY LIMITED 78,003,872,768.00   
17                      ACN 085 239 998 PTY LTD 39,085,239,998.00   
23  ADANI ABBOT POINT TERMINAL HOLDINGS PTY LTD 17,154,644,685.00   
24        ADCORP HOLDINGS AUSTRALIA PTY LIMITED 91,161,583,893.00   
25                      ADECCO HOLDINGS PTY LTD 11,003,652,088.00   
26                    ADELAIDE BRIGHTON LIMITED 15,007,596,018.00   
27                     ADIDAS AUSTRALIA PTY LTD 80,058,390,659.00   
28                ADITYA BIRLA MINERALS LIMITED 37,103,515,037.00   
29                ADM TRADING AUSTRALIA PTY LTD 92,126,806,979.00   
31             AEGIS AUSTRALIA HOLDINGS PTY LTD 60,147,301,700.00   

    Total income $  Taxable income $  Tax payable $  
0        452892659     86,587,682.00  25,328,872.00  
1       2561960532     91,824,776.00  22,270,429.00  
3        570890761     49,107,218.00  13,273,263.00  
5       1173330692      2,754,381.00            NaN  
6        400530322     24,939,732.00   7,481,920.00  
7        278129946     28,985,358.00   5,003,197.00  
9        210271588               NaN            NaN  
11      1409921569    170,420,860.00  50,502,197.00  
12       962674018      4,271,387.00   1,281,416.00  
15       622031079     27,455,363.00   7,829,730.00  
16       309078107               NaN            NaN  
17       479918758     32,039,735.00   9,611,920.00  
23       268376014               NaN            NaN  
24       296385800               NaN            NaN  
25       529972818               NaN            NaN  
26      1258243350    212,524,266.00  61,010,884.00  
27       233260318     14,458,827.00   4,337,648.00  
28       348620947     27,524,771.00     686,892.00  
29      1307023625               NaN            NaN  
31       388382443     28,409,118.00   4,729,182.00

In [8]:
#first year private dataset
df_private = pd.read_excel('2013-14-corporate-report-of-entity-tax-information.xlsx', sheet_name = 'March')
df_private

Name               ABN  \
0             21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
1                   A.C.N. 003 933 300 PTY LIMITED 58,003,933,300.00   
2                              AA HOLDINGS PTY LTD 94,112,233,744.00   
3                  ABC TISSUE PRODUCTS PTY LIMITED 33,003,085,112.00   
4    ACCOLADE WINES HOLDINGS AUSTRALIA PTY LIMITED 56,103,359,299.00   
..                                             ...               ...   
316          WORKFORCE INTERNATIONAL GROUP PTY LTD 29,141,197,097.00   
317                                WORKPAC PTY LTD 52,111,076,012.00   
318                             WSO CO PTY LIMITED 73,102,757,924.00   
319                  WT FINANCE (AUST) PTY LIMITED 16,108,806,711.00   
320                                XEGELAB PTY LTD 58,159,334,764.00   

     Total income $  Taxable income $  Tax payable $  
0         206773792               NaN            NaN  
1         338757483               NaN            NaN  
2         527619794     13,769,127.00   4,130,738.00  
3         416340995     54,388,372.00  16,316,433.00  
4         400115212               NaN            NaN  
..              ...               ...            ...  
316       203036205      4,429,373.00   1,328,812.00  
317       487342330               NaN            NaN  
318       237396548               NaN            NaN  
319      1429676509               NaN            NaN  
320       213431479               NaN            NaN  

[321 rows x 5 columns]

In [9]:
#filter each year of data to only include firms with $200M or more in total income
df_private = df_private[df_private['Total income $'] >= 200000000]
df_private.head(20)

Name               ABN  \
0            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
1                  A.C.N. 003 933 300 PTY LIMITED 58,003,933,300.00   
2                             AA HOLDINGS PTY LTD 94,112,233,744.00   
3                 ABC TISSUE PRODUCTS PTY LIMITED 33,003,085,112.00   
4   ACCOLADE WINES HOLDINGS AUSTRALIA PTY LIMITED 56,103,359,299.00   
5                  ADCO CONSTRUCTIONS PTY LIMITED 46,001,044,391.00   
6                                  ADCOME PTY LTD 99,110,315,681.00   
7                  AGGREGATE HOLDINGS PTY LIMITED 24,000,455,881.00   
8          AGNVET MANAGEMENT SERVICES PTY LIMITED 91,087,675,465.00   
9                            AHRENS GROUP PTY LTD 76,114,260,230.00   
10                          ALIMFRESH PTY LIMITED 78,002,831,843.00   
11                       ALINTA POWER CAT PTY LTD 63,124,509,619.00   
12             ALLIED MILLS AUSTRALIA PTY LIMITED 61,101,262,588.00   
13                 ALTOMONTE HOLDINGS PTY LIMITED 18,000,821,910.00   
14                  APCO SERVICE STATIONS PTY LTD 91,007,229,898.00   
15                       AQUASURE FINANCE PTY LTD 35,136,072,627.00   
16                      AQUASURE HOLDINGS PTY LTD 46,135,956,053.00   
17                              ARA GROUP PTY LTD 47,074,886,561.00   
18                     ARJ GROUP HOLDINGS PTY LTD 31,104,958,527.00   
19                  ARROW ENERGY HOLDINGS PTY LTD 94,141,385,293.00   

    Total income $  Taxable income $  Tax payable $  
0        206773792               NaN            NaN  
1        338757483               NaN            NaN  
2        527619794     13,769,127.00   4,130,738.00  
3        416340995     54,388,372.00  16,316,433.00  
4        400115212               NaN            NaN  
5        504172377      9,504,522.00   2,589,041.00  
6        240080911               NaN            NaN  
7        251925354     49,140,943.00   6,830,820.00  
8        265548816     12,866,045.00   3,504,100.00  
9        289628878     19,508,901.00   3,497,394.00  
10       205235396     39,654,158.00   3,135,469.00  
11      1698802352        449,839.00            NaN  
12       489474681     26,376,589.00   6,080,048.00  
13       399998346               NaN            NaN  
14       290762746      4,398,076.00   1,319,423.00  
15       355936371               NaN            NaN  
16       351754795               NaN            NaN  
17       301253004      7,783,667.00   2,328,507.00  
18       490763874     48,975,684.00  14,692,705.00  
19       292578318               NaN            NaN

In [10]:
# Do not keep any firms that are missing their Australian Business Number (ABN)
df_public.dropna(inplace = True)
df_public

/usr/local/lib/python3.7/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


Name               ABN  \
0                            3M AUSTRALIA PTY LTD 90,000,100,096.00   
1                              A P EAGERS LIMITED 87,009,680,013.00   
3                                    AAPC LIMITED 87,009,175,820.00   
6                  ABBOTT AUSTRALASIA PTY LIMITED 95,000,180,389.00   
7                              ABBVIE PTY LIMITED 48,156,384,262.00   
...                                           ...               ...   
1528             YARA PILBARA FERTILISERS PTY LTD 74,095,441,151.00   
1530               YOKOGAWA AUSTRALIA PTY LIMITED 36,003,888,364.00   
1534                ZF HOLDINGS AUSTRALIA PTY LTD 85,109,711,828.00   
1536                 ZOETIS AUSTRALIA PTY LIMITED 94,156,476,425.00   
1537  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   

      Total income $  Taxable income $  Tax payable $  
0          452892659     86,587,682.00  25,328,872.00  
1         2561960532     91,824,776.00  22,270,429.00  
3          570890761     49,107,218.00  13,273,263.00  
6          400530322     24,939,732.00   7,481,920.00  
7          278129946     28,985,358.00   5,003,197.00  
...              ...               ...            ...  
1528       424038887    221,091,145.00  66,327,344.00  
1530       220270930     28,495,683.00   8,517,716.00  
1534       204210113     18,799,144.00   4,662,674.00  
1536       256720802     16,509,304.00   4,952,791.00  
1537      2449944648    224,490,384.00  47,810,511.00  

[639 rows x 5 columns]

In [11]:
frames = [df_public, df_private]

df_public_private = pd.concat(frames, keys = ['public', 'private']).reset_index().drop('level_1', axis=1).rename(columns={'level_0':'Public/Private'})

df_public_private.insert(5, "Income Year", "13-14", True)
df_public_private

Public/Private                                   Name               ABN  \
0           public                   3M AUSTRALIA PTY LTD 90,000,100,096.00   
1           public                     A P EAGERS LIMITED 87,009,680,013.00   
2           public                           AAPC LIMITED 87,009,175,820.00   
3           public         ABBOTT AUSTRALASIA PTY LIMITED 95,000,180,389.00   
4           public                     ABBVIE PTY LIMITED 48,156,384,262.00   
..             ...                                    ...               ...   
955        private  WORKFORCE INTERNATIONAL GROUP PTY LTD 29,141,197,097.00   
956        private                        WORKPAC PTY LTD 52,111,076,012.00   
957        private                     WSO CO PTY LIMITED 73,102,757,924.00   
958        private          WT FINANCE (AUST) PTY LIMITED 16,108,806,711.00   
959        private                        XEGELAB PTY LTD 58,159,334,764.00   

     Total income $  Taxable income $ Income Year  Tax payable $  
0         452892659     86,587,682.00       13-14  25,328,872.00  
1        2561960532     91,824,776.00       13-14  22,270,429.00  
2         570890761     49,107,218.00       13-14  13,273,263.00  
3         400530322     24,939,732.00       13-14   7,481,920.00  
4         278129946     28,985,358.00       13-14   5,003,197.00  
..              ...               ...         ...            ...  
955       203036205      4,429,373.00       13-14   1,328,812.00  
956       487342330               NaN       13-14            NaN  
957       237396548               NaN       13-14            NaN  
958      1429676509               NaN       13-14            NaN  
959       213431479               NaN       13-14            NaN  

[960 rows x 7 columns]

### A- Continue to Assemble- Part 2

**Ultimately, you need to combine the seven datasets into one DataFrame. Before you start, think through what type of merge you want to do based on which observations you’d like to keep. Explain your reasoning.**

<div style="color:black;
           display:fill;
           border-radius:5px;
           background-color:#ABBAEA;>

            
<p style="padding: 10px;
              color:white;">

**Answer:**
We should merge these different dataframes by applying an inner join because all the dataframes have attributes that are in common such as "total income", "taxable "and "tax payable" as well as the names of the public and private companies correspond.  

----

**Transform the data. The steps below walk you through the data merging and cleaning process. There are different ways to get to the same end point. <font color = 'blue'> Follow the process below to demonstrate your skillset.**

**1.** Rename the columns in each dataset to distinguish which year each Total Income, Taxable Income and Taxes Payable come from.
- It is also important to make sure that the base of the column names are the same. For instance, if you make Total Income in year 1 `Total Income_14` make the second year’s Total Income column `Total Income_15`.
- This is most efficient if you write a loop but a loop is not required here. (Nor will you receive bonus points but you may save yourself some time and you will be better for it :))

In [12]:
# list of num files to clean
files = ['2014-15-corporate-report-of-entity-tax-information.xlsx', \
       '2015-16-corporate-report-of-entity-tax-information.xlsx', \
       '2016-17-corporate-report-of-entity-tax-information.xlsx', \
       '2017-18-corporate-report-of-entity-tax-information.xlsx', \
       '2018-19-corporate-report-of-entity-tax-information.xlsx', \
       '2019-20-corporate-report-of-entity-tax-information.xlsx']


# Total Compilation Variable (empty list to append cleaned dataset)
dict_df = {}

for file in files:
  # DATA LOADING (2014-15, 2015-16)
  if (file[0:7] == '2014-15' or file[0:7] == '2015-16'):
    year = file[0:7]

    # Load df from a sheet "(year)"
    df = pd.read_excel(file, sheet_name = year)
    
    # Create a new column "Income year" to df
    # Use forloop to insert file of "(year)" to "Income year" column
    df['Income year'] = year

  # DATA LOADING (2016-17 ... 2019-20)                
  else:
    # Load df from a sheet "Income tax details" 
    df = pd.read_excel(file, sheet_name = 1)

  # Final Merge: add "df" to "dict_df" (save each cleaned dataframe by appending to empty list)
  year = file[2:4]
  df = df[df['Total income $'] >= 200000000].reset_index(drop = True)
  df.dropna(subset=['ABN'], inplace =True)
  df = df.rename(columns={'Total income $' : 'Total income_' + year, 'Taxable income $' : 'Taxable income_' + year, 'Tax payable $' : 'Tax payable_' + year})
  dict_df[file[0:7]] = df
  

dict_df['2014-15']

Name               ABN  \
0            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
1                      318 RUSSELL STREET PTY LTD 33,146,526,052.00   
2                            3M AUSTRALIA PTY LTD 90,000,100,096.00   
3                              7 HOLDINGS PTY LTD 31,005,620,851.00   
4                              A P EAGERS LIMITED 87,009,680,013.00   
...                                           ...               ...   
1351             YARA PILBARA FERTILISERS PTY LTD 74,095,441,151.00   
1352               YOKOGAWA AUSTRALIA PTY LIMITED 36,003,888,364.00   
1353                        YOUI HOLDINGS PTY LTD 41,124,972,425.00   
1354                 ZOETIS AUSTRALIA PTY LIMITED 94,156,476,425.00   
1355  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   

      Total income_14  Taxable income_14  Tax payable_14 Income year  
0           216169773                NaN             NaN     2014-15  
1           244571771       2,879,632.00      863,890.00     2014-15  
2           451442784      78,533,075.00   22,740,986.00     2014-15  
3          2656137512     158,520,113.00   45,811,348.00     2014-15  
4          2690661986     148,269,844.00   34,422,220.00     2014-15  
...               ...                ...             ...         ...  
1351        375158366     135,867,563.00   40,760,269.00     2014-15  
1352        221570484      14,584,872.00    4,342,018.00     2014-15  
1353        579598994                NaN             NaN     2014-15  
1354        227923349      21,148,409.00    4,198,287.00     2014-15  
1355       2148575463     208,078,071.00   31,145,811.00     2014-15  

[1348 rows x 6 columns]

# New Section

**2.** For dataframes that disclose more than one year of data, filter out any late filings. i.e. Only keep data related to 2013-2014 that was disclosed in December of 2015 and same idea for all other years. Any data that is filed later will engender a later change in behavior, if any, which would make it harder to measure.

In [13]:
for year in dict_df.keys() : 
  dict_df[year] = dict_df[year][dict_df[year]['Income year'] == year]
  dict_df[year] = dict_df[year].drop('Income year', axis = 1)

**3.** Perform a wide merge in which each firm has only one row of data and subsequent years of information are presented in additional columns.

In [14]:
df_public_private = df_public_private.rename(columns={'Total income $' : 'Total income_' + '13',
                                                      'Taxable income $' : 'Taxable income_' + '13',
                                                      'Tax payable $' : 'Tax payable_' + '13'})
df_public_private = df_public_private.drop('Income Year', axis = 1)
df_public_private.head()

Public/Private                            Name               ABN  \
0         public            3M AUSTRALIA PTY LTD 90,000,100,096.00   
1         public              A P EAGERS LIMITED 87,009,680,013.00   
2         public                    AAPC LIMITED 87,009,175,820.00   
3         public  ABBOTT AUSTRALASIA PTY LIMITED 95,000,180,389.00   
4         public              ABBVIE PTY LIMITED 48,156,384,262.00   

   Total income_13  Taxable income_13  Tax payable_13  
0        452892659      86,587,682.00   25,328,872.00  
1       2561960532      91,824,776.00   22,270,429.00  
2        570890761      49,107,218.00   13,273,263.00  
3        400530322      24,939,732.00    7,481,920.00  
4        278129946      28,985,358.00    5,003,197.00

In [15]:
df_merge = df_public_private
for year in dict_df.keys() : 
  df_merge = pd.merge(df_merge, dict_df[year], how = 'left', on = ['Name', 'ABN'])

In [16]:
df_merge

Public/Private                                   Name               ABN  \
0           public                   3M AUSTRALIA PTY LTD 90,000,100,096.00   
1           public                     A P EAGERS LIMITED 87,009,680,013.00   
2           public                           AAPC LIMITED 87,009,175,820.00   
3           public         ABBOTT AUSTRALASIA PTY LIMITED 95,000,180,389.00   
4           public                     ABBVIE PTY LIMITED 48,156,384,262.00   
..             ...                                    ...               ...   
955        private  WORKFORCE INTERNATIONAL GROUP PTY LTD 29,141,197,097.00   
956        private                        WORKPAC PTY LTD 52,111,076,012.00   
957        private                     WSO CO PTY LIMITED 73,102,757,924.00   
958        private          WT FINANCE (AUST) PTY LIMITED 16,108,806,711.00   
959        private                        XEGELAB PTY LTD 58,159,334,764.00   

     Total income_13  Taxable income_13  Tax payable_13  Total income_14  \
0          452892659      86,587,682.00   25,328,872.00   451,442,784.00   
1         2561960532      91,824,776.00   22,270,429.00 2,690,661,986.00   
2          570890761      49,107,218.00   13,273,263.00   576,695,349.00   
3          400530322      24,939,732.00    7,481,920.00   408,006,805.00   
4          278129946      28,985,358.00    5,003,197.00   298,407,784.00   
..               ...                ...             ...              ...   
955        203036205       4,429,373.00    1,328,812.00              NaN   
956        487342330                NaN             NaN   632,124,347.00   
957        237396548                NaN             NaN              NaN   
958       1429676509                NaN             NaN              NaN   
959        213431479                NaN             NaN   295,382,090.00   

     Taxable income_14  Tax payable_14  Total income_15  ...  Tax payable_16  \
0        78,533,075.00   22,740,986.00   472,954,075.00  ...    3,063,687.00   
1       148,269,844.00   34,422,220.00 3,270,065,544.00  ...   24,300,076.00   
2        42,131,051.00   10,715,193.00   523,798,213.00  ...   12,079,549.00   
3        18,376,639.00    5,512,992.00   356,801,644.00  ...    6,563,884.00   
4        21,196,535.00    2,807,428.00              NaN  ...    3,805,470.00   
..                 ...             ...              ...  ...             ...   
955                NaN             NaN   234,982,225.00  ...    1,914,279.00   
956       6,807,102.00    2,042,131.00   733,440,981.00  ...    7,960,734.00   
957                NaN             NaN              NaN  ...             NaN   
958                NaN             NaN              NaN  ...             NaN   
959                NaN             NaN   338,666,357.00  ...             NaN   

     Total income_17  Taxable income_17  Tax payable_17  Total income_18  \
0     486,372,947.00      25,662,268.00    7,698,680.00   511,545,349.00   
1                NaN                NaN             NaN 4,029,416,057.00   
2     441,575,197.00       3,171,699.00             NaN 1,011,189,579.00   
3     349,006,641.00      34,366,127.00   10,309,838.00   386,490,370.00   
4     354,663,361.00      39,824,499.00    4,825,239.00   386,039,410.00   
..               ...                ...             ...              ...   
955              NaN                NaN             NaN              NaN   
956 1,322,158,086.00      29,099,912.00    8,729,974.00 1,411,398,697.00   
957              NaN                NaN             NaN              NaN   
958              NaN                NaN             NaN              NaN   
959              NaN                NaN             NaN              NaN   

     Taxable income_18  Tax payable_18  Total income_19  Taxable income_19  \
0        17,333,790.00    5,200,137.00   541,859,606.00      16,409,915.00   
1       122,528,679.00   27,143,529.00              NaN                NaN   
2        65,008,055.00   17,913,60

**4.** Remove any excess columns that we no longer need.

**5.**	Once the data is in a wide format and only includes the firms you want to keep for your analysis, reformat the data into a traditional panel data format (each row relates to one time period for one firm). Follow the following three steps:
1. Reformat the data to be extra long such that each row is a different value (Taxable Income, Tax Payable, or Total Income) for each firm for each year.
2. Create a column that indicates which time period each row relates to.
3. Reformat the data again to be slightly wider such that each firm-year is one row and there is a different column for Taxable Income, Tax Payable, and Total Income (in addition to name and ABN and filing status). Your data should look as if you stacked the original seven DataFrames on top of each other. Using this longer process ensures that you only keep the firms you want to keep for a proper analysis.


As a check point, your final DataFrame should have data relating to **154 unique private firms and 643 unique public firms**.

In [17]:
# Wide to extra long format
df_melt = pd.melt(df_merge, id_vars = ['Public/Private', 'Name', 'ABN'])
# Create column with time period
df_melt[['Metric', 'Income Year']] = df_melt['variable'].str.split('_', expand=True)
df_melt.drop('variable', axis = 1, inplace = True)
df_melt['Income Year'] = '20' + df_melt['Income Year']

# Reformat the data
df_reshaped = df_melt.pivot_table(index = ['Name', 'ABN', 'Public/Private', 'Income Year'], columns = 'Metric', values = 'value').reset_index()

df_reshaped

Metric                                         Name               ABN  \
0              21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
1              21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
2              21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
3              21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
4              21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
...                                             ...               ...   
4569    ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4570    ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4571    ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4572    ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4573    ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   

Metric Public/Private Income Year   Tax payable  Taxable income  \
0             private        2013           NaN             NaN   
1             private        2014           NaN             NaN   
2             private        2015  3,441,870.00   11,472,901.00   
3             private        2016  2,460,297.00    8,200,990.00   
4             private        2017  6,684,387.00   22,281,289.00   
...               ...         ...           ...             ...   
4569           public        2015 16,283,100.00  104,973,479.00   
4570           public        2016 25,735,708.00  136,982,422.00   
4571           public        2017 46,511,732.00  229,666,422.00   
4572           public        2018 59,389,437.00  271,817,164.00   
4573           public        2019           NaN             NaN   

Metric     Total income  
0        206,773,792.00  
1        216,169,773.00  
2        362,380,100.00  
3        253,742,807.00  
4        274,237,922.00  
...                 ...  
4569   1,379,106,272.00  
4570   1,541,150,451.00  
4571   2,124,355,799.00  
4572   2,957,379,565.00  
4573   4,229,115,978.00  

[4574 rows x 7 columns]

In [18]:
print("Number of unique public firms : ", str(df_reshaped[df_reshaped['Public/Private'] == 'public'].Name.nunique()))
print("Number of unique private firms : ", str(df_reshaped[df_reshaped['Public/Private'] == 'private'].Name.nunique()))

Number of unique public firms :  639
Number of unique private firms :  318


**6.	Once your final dataset is transformed, make sure that it is fully clean and ready for analysis.**

   **a.**	As noted on the first tab of the excel files, negative taxable incomes are not disclosed. Instead, this field is left blank. Similarly, tax liabilities of zero are not disclosed. For these firms, their tax payable field is left blank. Replace all taxable income and tax payable missing values with zero.


In [19]:
df_reshaped[['Tax payable', 'Taxable income']] = df_reshaped[['Tax payable', 'Taxable income']].fillna(0)
df_reshaped.columns.name = ''
df_reshaped

Name               ABN  \
0            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
1            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
2            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
3            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
4            21ST CENTURY RESORT HOLDINGS PTY LTD 33,104,201,014.00   
...                                           ...               ...   
4569  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4570  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4571  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4572  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   
4573  ZURICH FINANCIAL SERVICES AUSTRALIA LIMITED 11,008,423,372.00   

     Public/Private Income Year   Tax payable  Taxable income     Total income  
0           private        2013          0.00            0.00   206,773,792.00  
1           private        2014          0.00            0.00   216,169,773.00  
2           private        2015  3,441,870.00   11,472,901.00   362,380,100.00  
3           private        2016  2,460,297.00    8,200,990.00   253,742,807.00  
4           private        2017  6,684,387.00   22,281,289.00   274,237,922.00  
...             ...         ...           ...             ...              ...  
4569         public        2015 16,283,100.00  104,973,479.00 1,379,106,272.00  
4570         public        2016 25,735,708.00  136,982,422.00 1,541,150,451.00  
4571         public        2017 46,511,732.00  229,666,422.00 2,124,355,799.00  
4572         public        2018 59,389,437.00  271,817,164.00 2,957,379,565.00  
4573         public        2019          0.00            0.00 4,229,115,978.00  

[4574 rows x 7 columns]

**6b.**	Make sure all datatypes are appropriate. If not, adjust accordingly.

In [20]:
df_reshaped.dtypes


Name               object
ABN               float64
Public/Private     object
Income Year        object
Tax payable       float64
Taxable income    float64
Total income      float64
dtype: object

In [21]:
df_reshaped['ABN'] = df_reshaped['ABN'].astype('int') 
df_reshaped['Income Year'] = df_reshaped['Income Year'].astype('int') 

**6c.**	Take some time to understand the data you are working with. Are there any biases, errors or incomplete items that concern you? No need to explicitly answer that question but keep it in mind when performing the rest of your analysis. You can learn more about the data [here](https://www.ato.gov.au/Business/Large-business/In-detail/Tax-transparency/Corporate-tax-transparency-report-for-the-2014-15-income-year/).

In [22]:
df_reshaped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4574 entries, 0 to 4573
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Name            4574 non-null   object 
 1   ABN             4574 non-null   int64  
 2   Public/Private  4574 non-null   object 
 3   Income Year     4574 non-null   int64  
 4   Tax payable     4574 non-null   float64
 5   Taxable income  4574 non-null   float64
 6   Total income    4574 non-null   float64
dtypes: float64(3), int64(2), object(2)
memory usage: 250.3+ KB


**Export your data as a pickle file so that you can easily import it into part 2 of this case.**

In [23]:
# export cleaned data as a pickle file

# replace final with name of your dataframe
final = df_reshaped

# 'part1.p' is the name of the file that is exported
final.to_pickle('part1.p')

At a minimum, you must turn in your 3 questions before you can receive access to part 2 so that your questions are not influenced by the questions you will answer in Part 2. I am giving you leniency in how to manage your time but I highly recommend having your questions done and MOST of the cleaning done before class on 10/17. The whole case is due a week later on 10/24.